pkg/controller/infra/infra_controller.go

```diff
@@ -240,10 +530,97 @@ func (c *InfraController) syncCluster(key string) error {
 		return err
 	}
 
-	// Deep-copy otherwise we are mutating our cache.
-	// TODO: Deep-copy only when needed.
-	cluster = cluster.DeepCopy()
+	// Determine whether provisioning is needed
+	provisioning := clusterCondition(cluster, clusteroperator.ClusterHardwareProvisioning)
+	if provisioning != nil {
+		// For now, if we've already attempted provisioning, do not try again
+		// TODO: Figure out how provisioning should be retried and how backoff should
+		// be handled
+		if provisioning.Status == kapi.ConditionTrue {
+			return c.checkProvisioningJob(cluster)
+		}
+		return nil
+	}
+	provisioned := clusterCondition(cluster, clusteroperator.ClusterHardwareProvisioned)
+	if provisioned != nil && provisioned.Status == kapi.ConditionTrue {
+		return nil
+	}
+
+	return c.provisionCluster(cluster)
+}
+
+func (c *InfraController) checkProvisioningJob(cluster *clusteroperator.Cluster) error {
+	jobFound := true
+	if cluster.Status.ProvisioningJob != nil {
+		_, err := c.jobsLister.Jobs(cluster.Namespace).Get(cluster.Status.ProvisioningJob.Name)
+		if err != nil && errors.IsNotFound(err) {
+			jobFound = false
+		}
+	}
+	if cluster.Status.ProvisioningJob == nil || !jobFound {
+		setJobNotFoundStatus(cluster)
+		return c.updateClusterStatus(cluster)
+	}
+	return nil
+}
+
+func jobCondition(job *v1batch.Job, conditionType v1batch.JobConditionType) *v1batch.JobCondition {
+	for i, condition := range job.Status.Conditions {
+		if condition.Type == conditionType {
+			return &job.Status.Conditions[i]
+		}
+	}
+	return nil
+}
+
+func clusterCondition(cluster *clusteroperator.Cluster, conditionType clusteroperator.ClusterConditionType) *clusteroperator.ClusterCondition {
+	for i, condition := range cluster.Status.Conditions {
+		if condition.Type == conditionType {
+			return &cluster.Status.Conditions[i]
+		}
+	}
+	return nil
+}
 
+func setJobNotFoundStatus(cluster *clusteroperator.Cluster) {
```

Comment:
Here is another place where another condition would be helpful to indicate that the provisioning was attempted but failed.